packages/babel-parser/src/parser/expression.js

```diff
@@ -704,6 +699,75 @@ export default class ExpressionParser extends LValParser {
     }
   }
 
+  isSliceSimpleArgument(expr: N.Expression): boolean {
+    return (
+      (expr.type === "NumericLiteral" || expr.type === "Identifier") &&
+      expr.extra?.parenthesized !== true
+    );
+  }
+
+  checkSliceArgument(pos: number, expr: ?N.Expression) {
+    if (
+      !expr ||
+      this.isSliceSimpleArgument(expr) ||
+      (expr.type === "UnaryExpression" &&
+        expr.operator === "-" &&
+        expr.extra?.parenthesized !== true &&
+        this.isSliceSimpleArgument(expr.argument))
+    ) {
+      return;
+    }
+    this.raise(
+      pos,
+      "Slice argument requires to be either numeric literal or identifier",
+    );
+  }
+
+  parseSliceArgument(): ?N.Expression {
+    let expr = null;
+    this.eat(tt.colon);
```

Comment:
Shouldn't this be `this.expect`, or be inside an `if (this.eat(tt.colon))`?
Also please add a test where this expectation is not respected, e.g. `foo[1:2;3]`

Reply:
It would throw "unexpected token, expected `]`" now since `parseExpression` is not guarded by `tt.colon` check.